charmhelpers/contrib/openstack/context.py

```diff
@@ -444,7 +464,27 @@ def __call__(self):
         for rid in relation_ids(self.rel_name):
             self.related = True
             for unit in related_units(rid):
-                rdata = relation_get(rid=rid, unit=unit)
+                rdata = {}
+                # NOTE(jamespage):
+                # forwards compat with application data
+                # bag driven approach to relation.
+                _adata = relation_get(rid=rid, app=remote_service_name(rid))
+                if _adata:
+                    # New app data bag uses - instead of _
+                    # in key names - remap for compat with
+                    # existing relation data keys
+                    for key, value in _adata.items():
+                        if key == 'api-version':
+                            rdata[key.replace('-', '_')] = value.strip('v')
+                        else:
+                            rdata[key.replace('-', '_')] = value
+                    # Re-map some keys for backwards compatibility
+                    for target, source in self._forward_compat_remaps.items():
+                        rdata[target] = _adata.get(source)
```

Comment:
Is it worth wrapping this with a `try: except KeyError:` in case `_adata` doesn't contain a `source`?  i.e. is it always guaranteed, or do we want to raise the KeyError here if it doesn't?

Reply:
There's me being over-cautious.  I forgot that `dictionary.get()` supplies a `None`; python can be so inconsistent, as `.get_attribute` and friends all blow up!